### PR TITLE
feat: Add support for `abs`

### LINKED ISSUE #666: abs returns incorrect value in some cases
## SQL
```
SELECT c28, abs(c28) AS x FROM test1 ORDER BY c28;
```
### Spark Plan
```
AdaptiveSparkPlan isFinalPlan=true
+- == Final Plan ==
   *(2) Sort [c28#228L ASC NULLS FIRST], true, 0
   +- AQEShuffleRead coalesced
      +- ShuffleQueryStage 0
         +- Exchange rangepartitioning(c28#228L ASC NULLS FIRST, 200), ENSURE_REQUIREMENTS, [plan_id=24514]
            +- *(1) Project [c28#228L, abs(c28#228L) AS x#25310L]
               +- *(1) ColumnarToRow
                  +- FileScan parquet [c28#228L] Batched: true, DataFilters: [], Format: Parquet, Location: InMemoryFileIndex(1 paths)[file:/home/andy/git/apache/datafusion-comet/fuzz-testing/test1.parquet], PartitionFilters: [], PushedFilters: [], ReadSchema: struct<c28:bigint>
+- == Initial Plan ==
   Sort [c28#228L ASC NULLS FIRST], true, 0
   +- Exchange rangepartitioning(c28#228L ASC NULLS FIRST, 200), ENSURE_REQUIREMENTS, [plan_id=24500]
      +- Project [c28#228L, abs(c28#228L) AS x#25310L]
         +- FileScan parquet [c28#228L] Batched: true, DataFilters: [], Format: Parquet, Location: InMemoryFileIndex(1 paths)[file:/home/andy/git/apache/datafusion-comet/fuzz-testing/test1.parquet], PartitionFilters: [], PushedFilters: [], ReadSchema: struct<c28:bigint>

```
### Comet Plan
```
AdaptiveSparkPlan isFinalPlan=true
+- == Final Plan ==
   *(2) Sort [c28#228L ASC NULLS FIRST], true, 0
   +- AQEShuffleRead coalesced
      +- ShuffleQueryStage 0
         +- Exchange rangepartitioning(c28#228L ASC NULLS FIRST, 200), ENSURE_REQUIREMENTS, [plan_id=24579]
            +- *(1) ColumnarToRow
               +- CometProject [c28#228L, x#25316L], [c28#228L, abs(c28#228L) AS x#25316L]
                  +- CometScan parquet [c28#228L] Batched: true, DataFilters: [], Format: CometParquet, Location: InMemoryFileIndex(1 paths)[file:/home/andy/git/apache/datafusion-comet/fuzz-testing/test1.parquet], PartitionFilters: [], PushedFilters: [], ReadSchema: struct<c28:bigint>
+- == Initial Plan ==
   Sort [c28#228L ASC NULLS FIRST], true, 0
   +- Exchange rangepartitioning(c28#228L ASC NULLS FIRST, 200), ENSURE_REQUIREMENTS, [plan_id=24559]
      +- CometProject [c28#228L, x#25316L], [c28#228L, abs(c28#228L) AS x#25316L]
         +- CometScan parquet [c28#228L] Batched: true, DataFilters: [], Format: CometParquet, Location: InMemoryFileIndex(1 paths)[file:/home/andy/git/apache/datafusion-comet/fuzz-testing/test1.parquet], PartitionFilters: [], PushedFilters: [], ReadSchema: struct<c28:bigint>

```
First difference at row 11:
Spark: `-8779944534083813778,8779944534083813778`
Comet: `-8779944534083813778,-8779944534083813778`

### Steps to reproduce

_No response_

### Expected behavior

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by andygrove**
## Which issue does this PR close?

Closes #666 

## Rationale for this change

This PR is based on https://github.com/apache/datafusion-comet/pull/2595

## What changes are included in this PR?

The differences since https://github.com/apache/datafusion-comet/pull/2595 are:

- Revert test changes
- Implement new tests
- Fallback to Spark for unsupported types `DayTimeIntervalType` and `YearMonthIntervalType`
- Fix a potential bug with null scalar handling (although I am not sure if this code path even gets used with the Spark integration)

### CODE REVIEW

File: spark/src/test/scala/org/apache/comet/CometExpressionSuite.scala [hunk 1]
```diff
@@ -1430,74 +1428,6 @@ class CometExpressionSuite extends CometTestBase with AdaptiveSparkPlanHelper {
     testDoubleScalarExpr("expm1")
   }
 
-  // https://github.com/apache/datafusion-comet/issues/666
-  ignore("abs") {
```
Comment (andygrove) [inline spark/src/test/scala/org/apache/comet/CometExpressionSuite.scala:1434]: I implemented new tests using the fuzz data generator

**Comment by andygrove** [comment]
@hsiang-c could you review?

File: native/spark-expr/src/math_funcs/abs.rs [hunk 2]
```diff
@@ -0,0 +1,879 @@
...
+            dt => exec_err!("Not supported datatype for ABS: {dt}"),
+        },
+        ColumnarValue::Scalar(sv) => match sv {
+            ScalarValue::Null
+            | ScalarValue::UInt8(_)
+            | ScalarValue::UInt16(_)
+            | ScalarValue::UInt32(_)
+            | ScalarValue::UInt64(_) => Ok(args[0].clone()),
+            ScalarValue::Int8(a) => match a {
+                None => Ok(args[0].clone()),
```
Comment (hsiang-c) [inline native/spark-expr/src/math_funcs/abs.rs:202]: Good catch, thanks!

File: native/spark-expr/src/math_funcs/abs.rs [hunk 3]
```diff
@@ -0,0 +1,879 @@
...
+                }
+                _ => unreachable!(),
+            }
+        });
+    }
+
+    #[test]
+    fn test_abs_i8_scalar() {
+        with_fail_on_error(|fail_on_error| {
+            let args = ColumnarValue::Scalar(ScalarValue::Int8(Some(i8::MIN)));
```
Comment (hsiang-c) [inline native/spark-expr/src/math_funcs/abs.rs:349]: We might need a scalar test for the `None` case.

File: spark/src/main/scala/org/apache/comet/serde/math.scala [hunk 4]
```diff
@@ -144,6 +144,36 @@ object CometUnhex extends CometExpressionSerde[Unhex] with MathExprBase {
...
 }
 
+object CometAbs extends CometExpressionSerde[Abs] with MathExprBase {
+
+  override def getSupportLevel(expr: Abs): SupportLevel = {
+    expr.child.dataType match {
+      case _: NumericType =>
+        Compatible()
+      case _ =>
+        // Spark supports NumericType, DayTimeIntervalType, and YearMonthIntervalType
```
Comment (hsiang-c) [inline spark/src/main/scala/org/apache/comet/serde/math.scala:154]: LGTM

File: spark/src/test/scala/org/apache/comet/CometMathExpressionSuite.scala [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+
+  private def createTestData(generateNegativeZero: Boolean) = {
+    val r = new Random(42)
+    val schema = StructType(
+      Seq(
+        StructField("c0", DataTypes.ByteType, nullable = true),
+        StructField("c1", DataTypes.ShortType, nullable = true),
+        StructField("c2", DataTypes.IntegerType, nullable = true),
+        StructField("c3", DataTypes.LongType, nullable = true),
+        StructField("c4", DataTypes.FloatType, nullable = true),
```
Comment (hsiang-c) [inline spark/src/test/scala/org/apache/comet/CometMathExpressionSuite.scala:83]: Are `NaN`s included in the test data?

File: spark/src/test/scala/org/apache/comet/CometMathExpressionSuite.scala [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+  }
+
+  test("abs - negative zero") {
+    val df = createTestData(generateNegativeZero = true)
+    df.createOrReplaceTempView("tbl")
+    for (field <- df.schema.fields.filter(f =>
+        f.dataType == DataTypes.FloatType || f.dataType == DataTypes.DoubleType)) {
+      val col = field.name
+      checkSparkAnswerAndOperator(
+        s"SELECT $col, abs($col) FROM tbl WHERE signum($col) < 0 ORDER BY $col")
```
Comment (martin-g) [inline spark/src/test/scala/org/apache/comet/CometMathExpressionSuite.scala]: If the value of `$col` is `-0.0` (a negative zero) then `signum($col)` would return `0`.
Just checking whether this is the intended behavior because the test is about negative zero specifically.

File: native/spark-expr/src/math_funcs/abs.rs [hunk 7]
```diff
@@ -0,0 +1,879 @@
...
+            let args = ColumnarValue::Scalar(ScalarValue::UInt8(Some(u8::MAX)));
+            let fail_on_error_arg =
+                ColumnarValue::Scalar(ScalarValue::Boolean(Some(fail_on_error)));
+            match abs(&[args, fail_on_error_arg]) {
+                Ok(ColumnarValue::Scalar(ScalarValue::UInt8(Some(result)))) => {
+                    assert_eq!(result, u8::MAX);
+                    Ok(())
+                }
+                Err(e) => {
+                    if fail_on_error {
```
Comment (martin-g) [inline native/spark-expr/src/math_funcs/abs.rs]: Is it possible that it would overflow for unsigned integers ?
IMO it should always panic here, i.e. leave only the `else` clause body.

File: native/spark-expr/src/math_funcs/abs.rs [hunk 8]
```diff
@@ -0,0 +1,879 @@
...
+mod tests {
+    use super::*;
+    use datafusion::common::cast::{
+        as_decimal128_array, as_decimal256_array, as_float32_array, as_float64_array,
+        as_int16_array, as_int32_array, as_int64_array, as_int8_array, as_uint64_array,
+    };
+
+    fn with_fail_on_error<F: Fn(bool) -> Result<()>>(test_fn: F) {
+        for fail_on_error in [true, false] {
+            let _ = test_fn(fail_on_error);
```
Comment (martin-g) [inline native/spark-expr/src/math_funcs/abs.rs]: This would ignore the returned Result.

```suggestion
            let _ = test_fn(fail_on_error)?;
```

File: native/spark-expr/src/math_funcs/abs.rs [hunk 9]
```diff
@@ -0,0 +1,879 @@
...
+                    "The second argument must be boolean scalar, but got: {:?}",
+                    args[1]
+                );
+            }
+        }
+    } else {
+        false
+    };
+
+    match &args[0] {
```
Comment (martin-g) [inline native/spark-expr/src/math_funcs/abs.rs:86]: There is no check that the function is called with at least one argument.

File: native/spark-expr/src/math_funcs/abs.rs [hunk 10]
```diff
@@ -0,0 +1,879 @@
...
+                            // return the original value
+                            Ok(ColumnarValue::Scalar(ScalarValue::Int8(Some(*v))))
+                        } else {
+                            Err(arithmetic_overflow_error("Int8").into())
+                        }
+                    }
+                },
+            },
+            ScalarValue::Int16(a) => match a {
+                None => Ok(args[0].clone()),
```
Comment (martin-g) [inline native/spark-expr/src/math_funcs/abs.rs:216]: nit: The code for handling Int16/32/64 is very similar to Int8.
It could be extracted to a declarative macro.

File: spark/src/test/scala/org/apache/comet/CometMathExpressionSuite.scala (hunk 5 above)
Comment (andygrove) [reply spark/src/test/scala/org/apache/comet/CometMathExpressionSuite.scala:83]: Yes, the fuzz generator generates edge cases like `NaN`, `Infinity`, `-0.0` (optionally).

File: spark/src/test/scala/org/apache/comet/CometMathExpressionSuite.scala (hunk 6 above)
Comment (andygrove) [reply spark/src/test/scala/org/apache/comet/CometMathExpressionSuite.scala]: Thanks @martin-g. I updated the test. Results:

```
+----+-------+
|c4  |abs(c4)|
+----+-------+
|-0.0|0.0    |
|-0.0|0.0    |
|-0.0|0.0    |
|-0.0|0.0    |
|-0.0|0.0    |
|-0.0|0.0    |
|-0.0|0.0    |
|-0.0|0.0    |
|-0.0|0.0    |
|-0.0|0.0    |
|-0.0|0.0    |
|-0.0|0.0    |
|-0.0|0.0    |
|-0.0|0.0    |
|-0.0|0.0    |
|-0.0|0.0    |
|-0.0|0.0    |
|-0.0|0.0    |
|-0.0|0.0    |
|-0.0|0.0    |
+----+-------+
```

File: native/spark-expr/src/math_funcs/abs.rs (hunk 7 above)
Comment (andygrove) [reply native/spark-expr/src/math_funcs/abs.rs]: updated, thanks

File: native/spark-expr/src/math_funcs/abs.rs (hunk 9 above)
Comment (andygrove) [reply native/spark-expr/src/math_funcs/abs.rs:86]: added

File: native/spark-expr/src/math_funcs/abs.rs (hunk 8 above)
Comment (andygrove) [reply native/spark-expr/src/math_funcs/abs.rs]: Fixed. I couldn't add `?` but added `.expect`

File: native/spark-expr/src/math_funcs/abs.rs [hunk 11]
```diff
@@ -0,0 +1,871 @@
...
+            _ => Err(DataFusionError::Internal("Invalid data type".to_string())),
+        }
+    }};
+}
+
+/// This function mimics SparkSQL's [Abs]: https://github.com/apache/spark/blob/v4.0.1/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/arithmetic.scala#L148
+/// Spark's [ANSI-compliant]: https://spark.apache.org/docs/latest/sql-ref-ansi-compliance.html#arithmetic-operations dialect mode throws org.apache.spark.SparkArithmeticException
+/// when abs causes overflow.
+pub fn abs(args: &[ColumnarValue]) -> Result<ColumnarValue, DataFusionError> {
+    if args.is_empty() || args.len() > 2 {
```
Comment (comphead) [inline native/spark-expr/src/math_funcs/abs.rs:68]: I'm wondering why it can be 2 params, should it be a single param which is some numeric or interval?

Comment (comphead) [reply native/spark-expr/src/math_funcs/abs.rs:68]: https://spark.apache.org/docs/latest/api/sql/#abs

Only 1 expression in input

File: native/spark-expr/src/math_funcs/abs.rs [hunk 12]
```diff
@@ -0,0 +1,871 @@
...
+        return exec_err!("abs takes 1 or 2 arguments, but got: {}", args.len());
+    }
+
+    let fail_on_error = if args.len() == 2 {
+        match &args[1] {
+            ColumnarValue::Scalar(ScalarValue::Boolean(Some(fail_on_error))) => *fail_on_error,
+            _ => {
+                return exec_err!(
+                    "The second argument must be boolean scalar, but got: {:?}",
```
Comment (comphead) [inline native/spark-expr/src/math_funcs/abs.rs:77]: or Ic now, second param is ansi flag. In future we need to use `ScalarFunctionArgs.config_options` to get the flag, but it can be done in DF and then reused in Comet

File: native/spark-expr/src/math_funcs/abs.rs [hunk 13]
```diff
@@ -0,0 +1,871 @@
...
+            let args = ColumnarValue::Scalar(ScalarValue::UInt8(Some(u8::MAX)));
+            let fail_on_error_arg =
+                ColumnarValue::Scalar(ScalarValue::Boolean(Some(fail_on_error)));
+            match abs(&[args, fail_on_error_arg]) {
+                Ok(ColumnarValue::Scalar(ScalarValue::UInt8(Some(result)))) => {
+                    assert_eq!(result, u8::MAX);
+                    Ok(())
+                }
+                Err(e) => {
+                    panic!("Didn't expect error, but got: {e:?}")
```
Comment (comphead) [inline native/spark-expr/src/math_funcs/abs.rs]:
```suggestion
                    unreachable!("Didn't expect error, but got: {e:?}")
```
?

File: native/spark-expr/src/math_funcs/abs.rs [hunk 14]
```diff
@@ -0,0 +1,871 @@
...
+                }
+                Err(e) => {
+                    if fail_on_error {
+                        assert!(
+                            e.to_string().contains("ARITHMETIC_OVERFLOW"),
+                            "Error message did not match. Actual message: {e}"
+                        );
+                        Ok(())
+                    } else {
+                        panic!("Didn't expect error, but got: {e:?}")
```
Comment (comphead) [inline native/spark-expr/src/math_funcs/abs.rs]:
```suggestion
                        unreachable!("Didn't expect error, but got: {e:?}")
```

File: native/spark-expr/src/math_funcs/abs.rs [hunk 15]
```diff
@@ -0,0 +1,871 @@
...
```
Comment (comphead) [inline native/spark-expr/src/math_funcs/abs.rs]:
```suggestion
                        unreachable!("Didn't expect error, but got: {e:?}")
```

File: native/spark-expr/src/math_funcs/abs.rs [hunk 16]
```diff
@@ -0,0 +1,871 @@
...
+                }
+                _ => unreachable!(),
+            }
+        });
+    }
+
+    #[test]
+    fn test_abs_f32_array() {
+        with_fail_on_error(|fail_on_error| {
+            let input = Float32Array::from(vec![Some(-1f32), Some(f32::MIN), Some(f32::MAX), None]);
```
Comment (comphead) [inline native/spark-expr/src/math_funcs/abs.rs]: can we also have NaN, Infinities?

File: native/spark-expr/src/math_funcs/abs.rs (hunk 11 above)
Comment (andygrove) [reply native/spark-expr/src/math_funcs/abs.rs:68]: Yeah, this seems like a mistake. I checked Spark source and `Abs` is `UnaryExpression`

File: native/spark-expr/src/math_funcs/abs.rs (hunk 12 above)
Comment (andygrove) [reply native/spark-expr/src/math_funcs/abs.rs:77]: Ah, ok

File: native/spark-expr/src/math_funcs/abs.rs (hunk 16 above)
Comment (andygrove) [reply native/spark-expr/src/math_funcs/abs.rs]: added

File: native/spark-expr/src/math_funcs/abs.rs (hunk 11 above)
Comment (andygrove) [reply native/spark-expr/src/math_funcs/abs.rs:68]: nm, I see this is covered in the next comment

File: native/spark-expr/src/math_funcs/abs.rs [hunk 17]
```diff
@@ -0,0 +1,886 @@
...
+                        panic!("Didn't expect error, but got: {e:?}")
+                    }
+                }
+                _ => unreachable!(),
+            }
+        });
+    }
+
+    #[test]
+    fn test_abs_f32_array() {
```
Comment (comphead) [inline native/spark-expr/src/math_funcs/abs.rs:645]: Would be nice to have this edge case
```
scala> spark.sql("select abs(-0.0), abs(0.0)").show(false)
+--------+--------+
|abs(0.0)|abs(0.0)|
+--------+--------+
|0.0     |0.0     |
+--------+--------+
```

Because Rust 

```
fn main() {
    println!("{}", -0f32.abs());
}

-0
```

Comment (andygrove) [reply native/spark-expr/src/math_funcs/abs.rs:645]: Added. We also have specific Spark test for negative zero case.

**Review by comphead — APPROVED** [review]
Thanks @andygrove and @hsiang-c for supporting abs
